float64  current_x
